remotes/workqueue.go

```diff
@@ -0,0 +1,83 @@
+package remotes
+
+import (
+	"context"
+
+	"golang.org/x/sync/errgroup"
+)
+
+type workItem struct {
+	do   func(ctx context.Context) error
+	done chan struct{}
+	deps []chan struct{}
+}
+
+func newWorkItem(do func(ctx context.Context) error, deps ...chan struct{}) *workItem {
+	return &workItem{
+		do:   do,
+		done: make(chan struct{}),
+		deps: deps,
+	}
+}
+
+func (wi *workItem) process(ctx context.Context) error {
+	defer close(wi.done)
+	for _, d := range wi.deps {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-d:
+		}
+	}
+	select {
+	case <-ctx.Done():
+		return nil
+	default:
+		return wi.do(ctx)
+	}
+}
+
+type workQueue struct {
+	workGroup *errgroup.Group
+	todoList  chan *workItem
+	ctx       context.Context
+}
+
+//nolint: unparam
+func newWorkQueue(ctx context.Context, workerCount, todoBuffer int) *workQueue {
```

Comment:
I was hoping to find an existing implementation somewhere for this sort of thing. I did find https://godoc.org/k8s.io/client-go/util/workqueue (I was hoping more for something in golang.org/x/sync, oh well). Is it possible to use that?

Reply:
Having a look right now. The special thing we have here is "Work item dependencies" where some work items can begin only when a set of other work items have finished.

Reply:
Just had a look, there is no task dependencies support, and it feels like a very old part of the codebase, directly converted from the original k8s Java codebase. I really don't want to use that.

Reply:
That's fair. I had a bit of a look around and found ~dozen libraries (mostly via https://awesome-go.com/) but none of them had dependencies that I could see.